### PR TITLE
[@kadena/graph] Query, Object and Field descriptions

### PR DESCRIPTION
**Pull request by MRVDH**
Please check if all of the definitions of the queries, objects and the fields are correct.

### CODE REVIEW

File: packages/apps/graph/generated-schema.graphql [hunk 1]
```diff
@@ -3,8 +3,11 @@ The `BigInt` scalar type represents non-fractional signed whole numeric values.
 """
 scalar BigInt
 
+"""A unit of information that stores a set of verified transactions."""
```
Comment (nil-amrutlal-dept) [inline packages/apps/graph/generated-schema.graphql:6]: Would this also be true for an orphan block

File: packages/apps/graph/src/graph/objects/GraphConfiguration.ts [hunk 2]
```diff
@@ -1,12 +1,16 @@
 import { builder } from '../builder';
 
 export default builder.objectType('GraphConfiguration', {
-  description: 'General information about the graph and chainweb-data',
+  description: 'General information about the graph and chainweb-data.',
   fields: (t) => ({
-    maximumConfirmationDepth: t.exposeInt('maximumConfirmationDepth'),
+    maximumConfirmationDepth: t.exposeInt('maximumConfirmationDepth', {
+      description:
+        'The maximum confirmation depth configured in the environment variables.',
```
Comment (nil-amrutlal-dept) [inline packages/apps/graph/src/graph/objects/GraphConfiguration.ts:8]: same as above

File: packages/apps/graph/src/graph/objects/Block.ts [hunk 3]
```diff
@@ -99,6 +103,7 @@ export default builder.prismaNode('Block', {
           throw normalizeError(error);
         }
       },
+      description: 'The number of blocks that confirm this block.',
```
Comment (nil-amrutlal-dept) [inline packages/apps/graph/src/graph/objects/Block.ts:106]: same as above

File: packages/apps/graph/generated-schema.graphql [hunk 4]
```diff
@@ -31,6 +36,7 @@ type BlockTransactionsConnectionEdge {
   node: Transaction!
 }
 
+"""An account on a certain chain on a certain module, such as coin."""
```
Comment (alber70g) [inline packages/apps/graph/generated-schema.graphql:39]: _A chain specific fungible account_?

File: packages/apps/graph/generated-schema.graphql [hunk 5]
```diff
@@ -72,38 +78,55 @@ scalar DateTime
 """Floats that will have a value of 0 or more."""
 scalar Decimal
 
+"""A record of an execution of a function on a smart contract."""
```
Comment (alber70g) [inline packages/apps/graph/generated-schema.graphql:81]: _An event emitted by the execution of a smart-contract function_

File: packages/apps/graph/generated-schema.graphql [hunk 6]
```diff
@@ -72,38 +78,55 @@ scalar DateTime
...
 type Event implements Node {
   block: Block!
   chainId: BigInt!
+
+  """The block height of this event."""
```
Comment (alber70g) [inline packages/apps/graph/generated-schema.graphql:86]: _The height of the block where the event was emitted_

File: packages/apps/graph/generated-schema.graphql [hunk 7]
```diff
@@ -72,38 +78,55 @@ scalar DateTime
...
   moduleName: String!
   name: String!
+
+  """
+  The order index of this event, in the case that there are multiple events.
+  """
   orderIndex: BigInt!
   parameterText: String!
+
+  """The module name and the event name combined."""
```
Comment (alber70g) [inline packages/apps/graph/generated-schema.graphql:98]:
```suggestion
  """The full eventname, containing module and eventname, e.g. `coin.TRANSFER`"""
```

File: packages/apps/graph/generated-schema.graphql [hunk 8]
```diff
@@ -72,38 +78,55 @@ scalar DateTime
...
 
-"""General information about the graph and chainweb-data"""
+"""General information about the graph and chainweb-data."""
 type GraphConfiguration {
+  """
+  The maximum confirmation depth configured in the environment variables.
+  """
   maximumConfirmationDepth: Int!
+
+  """The minimum block height that is available in the graph."""
```
Comment (alber70g) [inline packages/apps/graph/generated-schema.graphql:111]:
```suggestion
  """The lowest block-height that is indexed in this endpoint"""
```

File: packages/apps/graph/generated-schema.graphql [hunk 9]
```diff
@@ -72,38 +78,55 @@ scalar DateTime
...
   minimumBlockHeight: BigInt
 }
 
+"""Access rule for a certain smart contract or account."""
```
Comment (alber70g) [inline packages/apps/graph/generated-schema.graphql:115]:
```suggestion
"""Guard for an account."""
```
I would ignore the smart-contract part for now. We use it only on accounts here.

File: packages/apps/graph/generated-schema.graphql [hunk 10]
```diff
@@ -72,38 +78,55 @@ scalar DateTime
...
 
+"""The public key of the miner that solved a block."""
 type MinerKey implements Node {
   block: Block!
   blockHash: String!
   id: ID!
   key: String!
 }
 
+"""An account on a certain module, such as coin."""
```
Comment (alber70g) [inline packages/apps/graph/generated-schema.graphql:129]:
```suggestion
"""A fungible-specific account"""
```

File: packages/apps/graph/generated-schema.graphql [hunk 11]
```diff
@@ -168,21 +191,50 @@ type PageInfo {
 scalar PositiveFloat
 
 type Query {
-  account(accountName: String!, moduleName: String!): ModuleAccount!
+  """Find an account by its name and its module, such as coin."""
```
Comment (alber70g) [inline packages/apps/graph/generated-schema.graphql:194]:
```suggestion
  """Retrieve an account by its name and fungible, such as coin."""
```

Here we also might want to change `moduleName` to `fungible`. Same for `ModuleAccount` -> `FungibleAccount`

File: packages/apps/graph/generated-schema.graphql [hunk 12]
```diff
@@ -168,21 +191,50 @@ type PageInfo {
...
+  account(accountName: String!, moduleName: String!): ModuleAccount
+
+  """Find a block by its hash."""
```
Comment (alber70g) [inline packages/apps/graph/generated-schema.graphql:197]:
```suggestion
  """Retrieve a block by hash."""
```
I dont know wheter all `find`s can be `retrieve`s. Find sounds more "voluntary". Bikeshedding though :)

File: packages/apps/graph/generated-schema.graphql [hunk 13]
```diff
@@ -168,21 +191,50 @@ type PageInfo {
...
   block(hash: String!): Block
+
+  """Find all blocks from a given height."""
   blocksFromHeight(chainIds: [String!], startHeight: Int!): [Block!]!
+
+  """
+  Find an account by its name and its module, such as coin, on a specific chain.
+  """
```
Comment (alber70g) [inline packages/apps/graph/generated-schema.graphql:205]:
```suggestion
  """
  Retrieve an account by its name, for a fungible, on a chain
  """
```

File: packages/apps/graph/generated-schema.graphql [hunk 14]
```diff
@@ -168,21 +191,50 @@ type PageInfo {
...
+  """Estimate the gas limit for a transaction."""
   gasLimitEstimate(transaction: PactTransaction!): Int!
+
+  """Estimate the gas limit for a list of transactions."""
   gasLimitEstimates(transactions: [PactTransaction!]!): [Int!]!
+
+  """Get the configuration of the graph."""
   graphConfiguration: GraphConfiguration!
+
+  """Get the height of the latest block."""
```
Comment (alber70g) [inline packages/apps/graph/generated-schema.graphql:220]:
```suggestion
  """Get the height of the block with the highest height."""
```
It's a subtle difference, but we don't retreive the `last` block height, but the highest block height

File: packages/apps/graph/generated-schema.graphql [hunk 15]
```diff
@@ -168,21 +191,50 @@ type PageInfo {
...
   lastBlockHeight: BigInt
   node(id: ID!): Node
   nodes(ids: [ID!]!): [Node]!
+
+  """Send a list of raw pact queries."""
```
Comment (alber70g) [inline packages/apps/graph/generated-schema.graphql:225]: Execute arbitrary Pact code via a `local` call without gas-estimation or signature-verification (e.g. `(+ 1 2)` or `(coin.get-details <account>`

File: packages/apps/graph/generated-schema.graphql [hunk 16]
```diff
@@ -168,21 +191,50 @@ type PageInfo {
...
   pactQueries(pactQuery: [PactQuery!]!): [String!]!
+
+  """Send a raw pact query."""
```
Comment (alber70g) [inline packages/apps/graph/generated-schema.graphql:228]: Idem

File: packages/apps/graph/generated-schema.graphql [hunk 17]
```diff
@@ -219,37 +271,61 @@ type QueryTransfersConnectionEdge {
   node: Transfer!
 }
 
+"""Information about a public key that can sign transactions."""
```
Comment (alber70g) [inline packages/apps/graph/generated-schema.graphql:274]:
```suggestion
"""A signer for a specific transaction"""
```

File: packages/apps/graph/generated-schema.graphql [hunk 18]
```diff
@@ -219,37 +271,61 @@ type QueryTransfersConnectionEdge {
...
+  """Subscribe to new blocks."""
   newBlocks(chainIds: [Int!]): [Block!]
+
+  """
+  Subscribe to a request key to wait for the related transaction to appear.
+  """
   transaction(requestKey: String!): Transaction
 }
 
+"""A request to execute a smart contract function."""
```
Comment (alber70g) [inline packages/apps/graph/generated-schema.graphql:301]:
```suggestion
"""A confirmed transaction."""
```

File: packages/apps/graph/generated-schema.graphql [hunk 19]
```diff
@@ -219,37 +271,61 @@ type QueryTransfersConnectionEdge {
...
 type Transaction implements Node {
+  """The JSON stringified error message if the transaction failed."""
   badResult: String
   block: Block
   chainId: BigInt!
+
+  """The PACT code that is executed."""
```
Comment (alber70g) [inline packages/apps/graph/generated-schema.graphql:308]:
```suggestion
  """
  The Pact expressions executed in this transaction when it's a `exec` transaction. 
  For a continuation, this field is `cont`
  """
```

File: packages/apps/graph/generated-schema.graphql [hunk 20]
```diff
@@ -219,37 +271,61 @@ type QueryTransfersConnectionEdge {
...
   code: String!
+
+  """
+  The JSON stringified continuation in the case that it is a continuation.
+  """
   continuation: String
   creationTime: DateTime!
+
+  """The JSON stringified data that is related to the request."""
```
Comment (alber70g) [inline packages/apps/graph/generated-schema.graphql:317]:
```suggestion
  """The `data` field of a transaction. Formatted as raw JSON"""
```

File: packages/apps/graph/generated-schema.graphql [hunk 21]
```diff
@@ -219,37 +271,61 @@ type QueryTransfersConnectionEdge {
...
   data: String
   eventCount: BigInt
   events: [Event!]
   gas: BigInt!
   gasLimit: BigInt!
   gasPrice: Float!
+
+  """The JSON stringified result if the transaction was successful."""
```
Comment (alber70g) [inline packages/apps/graph/generated-schema.graphql:325]:
```suggestion
  """The transaction result when it was successful. Formatted as raw JSON"""
```

File: packages/apps/graph/generated-schema.graphql [hunk 22]
```diff
@@ -261,25 +337,36 @@ type Transaction implements Node {
   rollback: Boolean
   senderAccount: String
   signers: [Signer!]
+
+  """The step number in the case that transactions are chained."""
```
Comment (alber70g) [inline packages/apps/graph/generated-schema.graphql:341]:
```suggestion
  """
  The step-number when this is a execution of a `defpact`, aka multi-step transaction.
  """
```

File: packages/apps/graph/generated-schema.graphql [hunk 23]
```diff
@@ -261,25 +337,36 @@ type Transaction implements Node {
...
 }
 
+"""A transfer of funds from a fungible between two accounts."""
 type Transfer implements Node {
   amount: Decimal!
   blockHash: String!
   blocks: [Block!]!
   chainId: BigInt!
+
+  """The transfer that is the counterparty of this transfer."""
```
Comment (alber70g) [inline packages/apps/graph/generated-schema.graphql:355]:
```suggestion
  """The counterpart of the crosschain-transfer. `null` when it's not a cross-chain-transfer"""
```

File: packages/apps/graph/generated-schema.graphql [hunk 24]
```diff
@@ -261,25 +337,36 @@ type Transaction implements Node {
...
   crossChainTransfer: Transfer
   height: BigInt!
   id: ID!
   moduleHash: String!
   moduleName: String!
+
+  """
+  The order of the transfer in the case that there are chained Transfers.
+  """
```
Comment (alber70g) [inline packages/apps/graph/generated-schema.graphql:364]:
```suggestion
  """
  The order of the transfer when it's a `defpact` execution.
  """
```

**Review by alber70g — COMMENTED** [review]
All comments are suggestion/other perspectives. Food for thought and discussion. Love the bikeshedding :D

File: packages/apps/graph/generated-schema.graphql (hunk 1 above)
Comment (MRVDH) [reply packages/apps/graph/generated-schema.graphql]: Something to review when we implement orphans I think

**Comment by alber70g** [comment]
> Two things I'd like to ask of the reviewers:
> 
> 1. Please check if all of the definitions of the queries, objects and the fields are correct.
> 2. Please help me find one-line explanations/descriptions of the following fields:
> 
> Block: -- epoch | What is the different with this and the creationTime field? Some epochs in the db are from 2019, and others from 2023.

block.epoch: The moment the difficulty is adjusted to maintain a block validation time of 30 seconds.

> MinerKey | Current description is "The public key of the miner that solved a block.", but I don't think that is correct, as a block can have multiple miner keys.

I think we should move the `minerkey` off of the `block` and put it to `blocks.miner`.

```
type Miner {
  keys: String[]
  predicate: String
  account: String
}
```

> 
> Signer | What is this specifically? Including the fields below.

In prisma.schema we have no relation to this table. But there is as there's a signer for each transaction. 

> -- address | Always empty, what is it? 

A transaction has a signer and a sender. A sender is an account, and the signer for the gas cap is the signer. A signer itself doesn't need an account. It just needs to be there as a sender. 

Also a signer signs for specific caps. Only the signer that signs for `coin.GAS` is associated with the sender_account_

> -- signature | Do we have an explanatory statement of this other than "the signature that was used to sign"?

The result of the signing operation of the hash of the transaction.

> 
> Transaction: 

> -- data | It seems the data can be pretty much anything. What is it specifically? 

Environment data made available to the transaction

> -- logs | This is always a value such as 'fBKTi_8_5T660fnFYvPTW9Wnbjj3276t95H95T9X55Y', but not sure what it means. 

Not sure if this is true though:

Identifier to retrieve the logs for the execution of the transaction

> -- metadata | Always empty, what is it? 

No clue

> -- pactId | A value such as 'MLfYVl8o6D70PzX1H1p1YRwnNGYNp2VHZPyL6kbKctA', what is it? 



> -- proof | A base64 encoded string with an object, input and subject. How should we describe this? 



> -- rollback | Not sure what this is.

`pactId` `proof` and `rollback` have to do with `defpacts` (multistep transactions)

`proof` is usually done when a crosschain tx is done, as the counterpart needs the proof to be able to execute

**Comment by alber70g** [comment]
> > > -- pactId | A value such as 'MLfYVl8o6D70PzX1H1p1YRwnNGYNp2VHZPyL6kbKctA', what is it?
> > 
> > 
> > The value of `pactId` is only filled when it is a cross-chain transaction. The `pactId` is the `requestKey` of the starting transaction. So in cross-chain transactions we always have a starting transaction (on the origin chain) and a finishing transaction (on the target chain). The `pactId` will be filled on the finishing transaction and it will have the value of the starting transaction's `requestKey`
> 
> Wouldn't it be more logical to call the field `crossChainTransactionRequestKey`? cc: @alber70g Or is there a specific reason it's called "pactId"?

Yes, it's for any execution of a `defpact`. So not just cross-chain transactions, also multi-step transactions

**Comment by MRVDH** [comment]
- Added descriptions everywhere
- Renamed module to fungible
- Removed miner keys and replaced it with miner account
- Added a max take on unlimited prisma calls

**Review by nil-amrutlal-dept — APPROVED** [review]

**Pull request merged**